chain/client/src/chunk_validation.rs

```diff
@@ -0,0 +1,153 @@
+use near_network::types::{NetworkRequests, PeerManagerMessageRequest};
+use near_primitives::checked_feature;
+use near_primitives::chunk_validation::{
+    ChunkEndorsement, ChunkEndorsementInner, ChunkEndorsementMessage, ChunkStateWitness,
+};
+use near_primitives::sharding::ShardChunkHeader;
+use near_primitives::types::EpochId;
+use near_primitives::validator_signer::ValidatorSigner;
+
+use crate::Client;
+
+/// A module that handles chunk validation logic. Chunk validation refers to a
+/// critical process of stateless validation, where chunk validators (certain
+/// validators selected to validate the chunk) verify that the chunk's state
+/// witness is correct, and then sends chunk endorsements to the block producer
+/// so that the chunk can be included in the block.
+pub struct ChunkValidator {
+    my_signer: Option<Arc<dyn ValidatorSigner>>,
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    network_sender: Sender<PeerManagerMessageRequest>,
+    runtime_adapter: Arc<dyn RuntimeAdapter>,
+}
+
+impl ChunkValidator {
+    pub fn new(
+        my_signer: Option<Arc<dyn ValidatorSigner>>,
+        epoch_manager: Arc<dyn EpochManagerAdapter>,
+        network_sender: Sender<PeerManagerMessageRequest>,
+        runtime_adapter: Arc<dyn RuntimeAdapter>,
+    ) -> Self {
+        Self { my_signer, epoch_manager, network_sender, runtime_adapter }
+    }
+
+    /// Performs the chunk validation logic. When done, it will send the chunk
+    /// endorsement message to the block producer. The actual validation logic
+    /// happens in a separate thread.
+    pub fn start_validating_chunk(&self, state_witness: ChunkStateWitness) -> Result<(), Error> {
+        if self.my_signer.is_none() {
+            return Err(Error::NotAValidator);
+        }
+        let chunk_header = state_witness.chunk_header.clone();
+        let epoch_id =
+            self.epoch_manager.get_epoch_id_from_prev_block(chunk_header.prev_block_hash())?;
+        // We will only validate something if we are a chunk validator for this chunk.
+        // Note this also covers the case before the protocol upgrade for chunk validators,
+        // because the chunk validators will be empty.
+        let chunk_validators = self.epoch_manager.get_chunk_validators(
+            &epoch_id,
+            chunk_header.shard_id(),
+            chunk_header.height_created(),
+        )?;
+        if !chunk_validators.contains_key(self.my_signer.as_ref().unwrap().validator_id()) {
+            return Err(Error::NotAChunkValidator);
+        }
+        let block_producer =
+            self.epoch_manager.get_block_producer(&epoch_id, chunk_header.height_created())?;
+
+        let network_sender = self.network_sender.clone();
+        let signer = self.my_signer.clone().unwrap();
+        let runtime_adapter = self.runtime_adapter.clone();
+        rayon::spawn(move || match validate_chunk(&state_witness, &*runtime_adapter) {
+            Ok(()) => {
+                tracing::debug!(
```

Comment:
nit: Having chunk hash and block producer not as a part of the message should make logs easier to work with. Also that seems to be a bit more common pattern in our code base. So maybe consider moving those outside of the message.

Reply:
yeah, having separate fields instead of string interpolation into the message